### PR TITLE
Fixes #6631: WCF method with UriTemplate annotation fails with InvalidOperationException

### LINKED ISSUE #6631: WCF method with UriTemplate annotation fails with InvalidOperationException
## Description

I have an existing WCF service that was originally developed in .NET/Windows, which I've been trying to get running on Mono/Mac/XSP. Everything works well, except for methods that use path parameters in the request URL through UriTemplate. 

When such method is called with Mono, the service returns a 500 error due to `System.InvalidOperationException: HttpListenerContext does not match any of the registered channels`; however, the same call works on .NET

The issue appears to be specific to using path parameters with UriTemplate, as methods that do NOT use UriTemplate, work equally well with Mono or .NET

**This issue has been previously reported in 2014 in the following [Bugzilla bug #25444](https://bugzilla.xamarin.com/show_bug.cgi?id=25444)**

## Steps to Reproduce

1. Create a simple WCF service that uses a UriTemplate with at least one path parameter on one of the OperationContract's
2. Call the service and observe the error

See the following sample to reproduce the issue:
[Sample Service.zip](https://github.com/mono/mono/files/1653492/Sample.Service.zip)

## Current Behavior

Service responds with a 500 status due to an exception during the processing of the request. 

screen shot 2018-01-22 at 3 31 12 pm

## Expected Behavior

The service should return the requested data and respond with a successful 200 status code

### On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

5.8.0

### Stacktrace

```
at System.ServiceModel.Channels.Http.HttpListenerManager.ProcessNewContext (System.ServiceModel.Channels.Http.HttpContextInfo ctxi) [0x0000a] in /Users/dtychshenko/dev/mono/mcs/class/System.ServiceModel/System.ServiceModel.Channels.Http/HttpListenerManager.cs:80 
  at System.ServiceModel.Channels.SvcHttpHandler.ProcessRequest (System.Web.HttpContext context) [0x00055] in /Users/dtychshenko/dev/mono/mcs/class/System.ServiceModel/System.ServiceModel.Channels/SvcHttpHandler.cs:83 
  at System.Web.HttpApplication+<Pipeline>d__225.MoveNext () [0x008fd] in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Web/System.Web/HttpApplication.cs:1338 
  at System.Web.HttpApplication.Tick () [0x00000] in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Web/System.Web/HttpApplication.cs:927 ```

### PR DESCRIPTION
**Pull request by dtychshenko**
As the title states, fixes #6631, which mirrors [Bugzilla bug #25444](https://bugzilla.xamarin.com/show_bug.cgi?id=25444)

This adds support to WCF services that use UriTemplate annotation with path parameters. 

For example, if a WCF service is hosted at endpoint:
```http://localhost/MyService.svc```
and has an operation contract for
```[WebGet(UriTemplate = "Books/{id}")]```
when a GET request is made to:
```http://localhost/MyServcie.svc/Books/2```
the HttpListenerManagerTable would be able to retrieve the existing listener for `/MyService.svc` as the request URL starts with `/MyService.svc`

### CODE REVIEW

**Comment by marek-safar** [comment]
approve

File: mcs/class/System.ServiceModel/System.ServiceModel.Channels.Http/HttpListenerManagerTable.cs [hunk 1]
```diff
@@ -107,6 +107,12 @@ public HttpListenerManager GetOrCreateManager (Uri uri, HttpTransportBindingElem
 					if (m != null)
 						return m;
 				}
+
+				// Lastly, try to match the listener to the start of the current request path 
+				// This is to support WCF methods with path parameters in UriTemplate annotation
+				m = listeners.FirstOrDefault (p => absolutePath.StartsWith (p.Key.AbsolutePath)).Value;
```
Comment (marek-safar) [inline mcs/class/System.ServiceModel/System.ServiceModel.Channels.Http/HttpListenerManagerTable.cs]: Should not the path be terminated with `/` ?
Please use StringComparison.Ordinal argument

Comment (dtychshenko) [reply mcs/class/System.ServiceModel/System.ServiceModel.Channels.Http/HttpListenerManagerTable.cs]: Not sure what you mean by the terminating `/`. The listener path would not have an ending `/`. The request path may or may not have an ending `/`, but both cases are supported by this.

Good point on the ordinal comparison 👍  I've added the argument.

**Comment by dtychshenko** [comment]
Any updates on this?

**Pull request merged**